File: AI_Model_testing/ML/Computational_Intelligence/CI_Model/PSO/Core/Engine.cpp
// Engine.cpp ส่วนหลักของการคำนวน pso
#include "Engine.hpp"
#include "Fitness.hpp"
#include <random>
#include <cmath>

using namespace std;

EnginePSO::EnginePSO(int particlesB, int iterations, const vector<vector<double>> &data , const vector<int> &y_data ,int num_classes)
    : particles_(particlesB),
      iterations_(iterations),
      X(data),
      y(y_data),
      num_class(num_classes),
      n_features(data[0].size()),
      particles(particlesB, vector<int>(n_features)),
      velocity(particlesB, vector<double>(n_features)),
      pbest(particlesB, vector<int>(n_features)),
      pbest_score(particlesB),
      gbest(n_features),
      gbest_score(-std::numeric_limits<double>::infinity())
{}



  EnginePSO:: PSOResult EnginePSO::run_parallel(int n_threads)
{
    initialize();

    int block = particles_ / n_threads;
    int remainder = particles_ % n_threads;

    vector<thread> threads;

    for (int i = 0; i < n_threads; i++)
    {
        int start = i * block + min(i, remainder);
        int end = start + block;

        if (i < remainder)
            end++;

        threads.emplace_back(
            &EnginePSO::IterationLoop,
            this,
            start,
            end
        );
    }

    for (auto &t : threads)
        t.join();

    PSOResult result;
    result.best_features = gbest;
    result.best_score = gbest_score.load();

    return result;
}

void EnginePSO::IterationLoop(int start, int end)
{
    int patience = 20;
    int no_improve_count = 0;
    double local_best_known = gbest_score.load();

    for (int iter = 0; iter < iterations_; iter++)
    {
        vector<int> gbest_snapshot = gbest;
        
        update_velocity(start, end, gbest_snapshot);
        update_position(start, end);

        for (int i = start; i < end; i++)
        {

            double score ;
            if (num_class == 2) {
                score = evaluate_binary(particles[i], X, y); 
            } else {
                score = evaluate_multiclass(particles[i], X, y, num_class); 
            }

            if (score > pbest_score[i])
            {
                pbest_score[i] = score;
                pbest[i] = particles[i];
            }
            if (score > gbest_score.load())
            {
                lock_guard<mutex> lock(mtx); 
                
                if (score > gbest_score.load()) 
                {
                    gbest_score.store(score);
                    gbest = particles[i];
                }
            }
        }
        double current_global_best = gbest_score.load();

        if (current_global_best > local_best_known + 1e-9) 
        {
            local_best_known = current_global_best;
            no_improve_count = 0; 
        } 

        else no_improve_count++;   

        if (no_improve_count >= patience) break; 
        
    }
}

void EnginePSO::initialize()
{
    mt19937 gen(random_device{}());
    bernoulli_distribution binary(0.5);
    uniform_real_distribution<double> vel(-1, 1);

    gbest_score.store(-std::numeric_limits<double>::infinity());

    for (int i = 0; i < particles_; i++)
    {
        for (int j = 0; j < n_features; j++)
        {
            particles[i][j] = binary(gen);
            velocity[i][j] = vel(gen);
        }

        pbest[i] = particles[i];

        double score ;
        if (num_class == 2) {
            score = evaluate_binary(particles[i], X, y); 
        } else {
            score = evaluate_multiclass(particles[i], X, y, num_class); 
        }
        
        pbest_score[i] = score;

        if (score > gbest_score.load())
        {
            gbest_score.store(score);
            gbest = particles[i];
        }
    }
}

void EnginePSO::update_velocity(int start, int end, const vector<int> &gbest_snapshot)
{
    double w = 0.7;
    double c1 = 1.5;
    double c2 = 1.5;

    thread_local mt19937 gen(std::random_device{}() + std::hash<std::thread::id>{}(std::this_thread::get_id()));
    uniform_real_distribution<double> rand01(0, 1);

    for (int i = start; i < end; i++)
    {
        for (int j = 0; j < n_features; j++)
        {
            double r1 = rand01(gen);
            double r2 = rand01(gen);

            velocity[i][j] =
                w * velocity[i][j] + c1 * r1 * (pbest[i][j] - particles[i][j]) + c2 * r2 * (gbest_snapshot[j] - particles[i][j]);
        }
    }
}

void EnginePSO::update_position(int start, int end)
{
    thread_local mt19937 gen(std::random_device{}() + std::hash<std::thread::id>{}(std::this_thread::get_id()));
    uniform_real_distribution<double> rand01(0, 1);

    for (int i = start; i < end; i++)
    {
        for (int j = 0; j < n_features; j++)
        {
            double s = 1.0 / (1.0 + std::exp(-velocity[i][j]));

            particles[i][j] =
                (rand01(gen) < s) ? 1 : 0;
        }
    }
}

File: AI_Model_testing/ML/Computational_Intelligence/CI_Model/PSO/Python_PSO/Core/Engine.py
# Engine.py - Python equivalent of Engine.cpp
# Main calculation engine for Binary Particle Swarm Optimization (BPSO)

import math
import random
from .Fitness import evaluate_binary, evaluate_multiclass


class EnginePSO:
    """
    Exact 1:1 translation of C++ EnginePSO in Engine.cpp
    """
    def __init__(self, particles_count, iterations, data, y_data, num_classes):
        self.particles_ = particles_count
        self.iterations_ = iterations
        self.X = data
        self.y = y_data
        self.num_class = num_classes
        self.n_features = len(data[0])

        self.particles = [[0] * self.n_features for _ in range(self.particles_)]
        self.velocity = [[0.0] * self.n_features for _ in range(self.particles_)]

        self.pbest = [[0] * self.n_features for _ in range(self.particles_)]
        self.pbest_score = [-float('inf')] * self.particles_

        self.gbest = [0] * self.n_features
        self.gbest_score = -float('inf')

    def initialize(self):
        """
        Initialize particles with Bernoulli(0.5) and velocity with Uniform(-1, 1)
        """
        self.gbest_score = -float('inf')

        for i in range(self.particles_):
            for j in range(self.n_features):
                self.particles[i][j] = 1 if random.random() < 0.5 else 0
                self.velocity[i][j] = random.uniform(-1.0, 1.0)

            self.pbest[i] = list(self.particles[i])

            if self.num_class == 2:
                score = evaluate_binary(self.particles[i], self.X, self.y)
            else:
                score = evaluate_multiclass(self.particles[i], self.X, self.y, self.num_class)

            self.pbest_score[i] = score

            if score > self.gbest_score:
                self.gbest_score = score
                self.gbest = list(self.particles[i])

    def update_velocity(self, gbest_snapshot):
        """
        Update particle velocities using standard PSO formula:
        v = w*v + c1*r1*(pbest - x) + c2*r2*(gbest - x)
        """
        w = 0.7
        c1 = 1.5
        c2 = 1.5

        for i in range(self.particles_):
            for j in range(self.n_features):
                r1 = random.random()
                r2 = random.random()
                self.velocity[i][j] = (
                    w * self.velocity[i][j]
                    + c1 * r1 * (self.pbest[i][j] - self.particles[i][j])
                    + c2 * r2 * (gbest_snapshot[j] - self.particles[i][j])
                )

    def update_position(self):
        """
        Update particle binary positions using Sigmoid transfer function:
        s(v) = 1 / (1 + exp(-v))
        x = 1 if rand() < s(v) else 0
        """
        for i in range(self.particles_):
            for j in range(self.n_features):
                s = 1.0 / (1.0 + math.exp(-self.velocity[i][j]))
                self.particles[i][j] = 1 if random.random() < s else 0

    def run(self):
        """
        Main optimization loop with early stopping patience of 20 iterations
        """
        self.initialize()

        patience = 20
        no_improve_count = 0
        local_best_known = self.gbest_score

        for _ in range(self.iterations_):
            gbest_snapshot = list(self.gbest)

            self.update_velocity(gbest_snapshot)
            self.update_position()

            for i in range(self.particles_):
                if self.num_class == 2:
                    score = evaluate_binary(self.particles[i], self.X, self.y)
                else:
                    score = evaluate_multiclass(self.particles[i], self.X, self.y, self.num_class)

                if score > self.pbest_score[i]:
                    self.pbest_score[i] = score
                    self.pbest[i] = list(self.particles[i])

                if score > self.gbest_score:
                    self.gbest_score = score
                    self.gbest = list(self.particles[i])

            current_global_best = self.gbest_score

            if current_global_best > local_best_known + 1e-9:
                local_best_known = current_global_best
                no_improve_count = 0
            else:
                no_improve_count += 1

            if no_improve_count >= patience:
                break

        return self.gbest, self.gbest_score
